File: packages/skill-eval/src/report/console-reporter.ts
import pc from "picocolors";
import type { OrchestratorResult } from "../runner/orchestrator.js";
import type { EvalCase } from "../schemas/evals.js";
import { getTurns } from "../schemas/evals.js";
import { rateColor } from "../utils/rate.js";

const LABEL_TRUNCATE = 100;
const EVIDENCE_TRUNCATE = 200;

export function printResults(result: OrchestratorResult, evals: EvalCase[], providerCount: number) {
	console.log(pc.bold("\n  Results\n"));
	const byEval = new Map<number, typeof result.runs>();
	for (const run of result.runs) {
		if (!byEval.has(run.eval_id)) byEval.set(run.eval_id, []);
		byEval.get(run.eval_id)?.push(run);
	}

	let totalPassed = 0;
	let totalFailed = 0;

	for (const [evalId, runs] of byEval) {
		const evalCase = evals.find((e) => e.id === evalId);
		const turns = evalCase ? getTurns(evalCase) : [];
		const label = (turns[0] ?? "").slice(0, LABEL_TRUNCATE);
		console.log(
			`  ${pc.bold(`Eval ${evalId}`)} ${pc.dim("—")} ${label}${label.length >= LABEL_TRUNCATE ? "..." : ""}`,
		);

		for (const run of runs) {
			const r = run.result;
			totalPassed += r.passed;
			totalFailed += r.failed;

			const providerLabel = providerCount > 1 ? `${pc.dim("[")}${run.model}${pc.dim("]")} ` : "";

			for (const exp of run.expectations) {
				if (exp.passed) {
					console.log(`    ${pc.green("✓")} ${providerLabel}${exp.text}`);
				} else {
					console.log(`    ${pc.red("✗")} ${providerLabel}${exp.text}`);
					if (exp.evidence) {
						console.log(`      ${pc.dim(exp.evidence.slice(0, EVIDENCE_TRUNCATE))}`);
					}
				}
			}

			if (run.eval_feedback) {
				console.log(`    ${pc.dim(`› ${run.eval_feedback}`)}`);
			}

			if (providerCount > 1) {
				console.log(
					`    ${pc.dim(`${r.passed}/${r.total} passed · ${r.time_seconds.toFixed(1)}s`)}`,
				);
			}
		}
		console.log("");
	}

	const totalAssertions = totalPassed + totalFailed;
	const overallRate = totalAssertions > 0 ? totalPassed / totalAssertions : 0;
	const scoreColor = rateColor(overallRate);

	console.log(pc.bold("  Scorecard\n"));
	console.log(
		`  ${scoreColor(pc.bold(`${(overallRate * 100).toFixed(0)}%`))} overall  ${pc.dim(`(${totalPassed} passed, ${totalFailed} failed out of ${totalAssertions} assertions)`)}`,
	);

	for (const [key, stats] of Object.entries(result.providerSummary)) {
		const color = rateColor(stats.pass_rate.mean);
		console.log(
			`  ${color(pc.bold(`${(stats.pass_rate.mean * 100).toFixed(0)}%`))} ${key}` +
				`  ${pc.dim(`${stats.time_seconds.mean.toFixed(1)}s avg · ${stats.total_tokens.mean.toFixed(0)} tok · $${stats.cost_usd.mean.toFixed(4)}`)}`,
		);
	}
	console.log("");
}

File: packages/skill-eval-python/src/skill_eval/report/console_reporter.py
from __future__ import annotations

from rich.console import Console

from skill_eval.schemas.evals import EvalCase, get_turns
from skill_eval.utils.rate import rate_color

LABEL_TRUNCATE = 100
EVIDENCE_TRUNCATE = 200


def print_results(
    result: dict,
    evals: list[EvalCase],
    provider_count: int,
) -> None:
    console = Console()
    console.print("\n  [bold]Results[/bold]\n")

    by_eval: dict[int, list] = {}
    for run in result["runs"]:
        eval_id = run.get("eval_id") or run.eval_id
        if eval_id not in by_eval:
            by_eval[eval_id] = []
        by_eval[eval_id].append(run)

    total_passed = 0
    total_failed = 0

    for eval_id, runs in sorted(by_eval.items()):
        eval_case = next((e for e in evals if e.id == eval_id), None)
        turns = get_turns(eval_case) if eval_case else []
        label = (turns[0] or "")[:LABEL_TRUNCATE]
        suffix = "..." if len(label) >= LABEL_TRUNCATE else ""
        console.print(f"  [bold]Eval {eval_id}[/bold] [dim]—[/dim] {label}{suffix}")

        for run in runs:
            r = run.get("result") or run.result
            total_passed += r.get("passed", r.passed)
            total_failed += r.get("failed", r.failed)

            provider_label = ""
            if provider_count > 1:
                model = run.get("model") or run.model
                provider_label = f"[dim][[/dim]{model}[dim]][/dim] "

            for exp in run.get("expectations") or run.expectations:
                text = exp.get("text") or exp.text
                passed = exp.get("passed", exp.passed)
                if passed:
                    console.print(f"    [green]✓[/green] {provider_label}{text}")
                else:
                    console.print(f"    [red]✗[/red] {provider_label}{text}")
                    evidence = exp.get("evidence", getattr(exp, "evidence", "")) or ""
                    if evidence:
                        console.print(
                            f"      [dim]{evidence[:EVIDENCE_TRUNCATE]}[/dim]"
                        )

            eval_feedback = run.get("eval_feedback") or getattr(
                run, "eval_feedback", None
            )
            if eval_feedback:
                console.print(f"    [dim]› {eval_feedback}[/dim]")

            if provider_count > 1:
                passed_val = r.get("passed", r.passed)
                total_val = r.get("total", r.total)
                time_val = r.get("time_seconds", r.time_seconds)
                console.print(
                    f"    [dim]{passed_val}/{total_val} passed · "
                    f"{time_val:.1f}s[/dim]"
                )
        console.print()

    total_assertions = total_passed + total_failed
    overall_rate = total_passed / total_assertions if total_assertions > 0 else 0
    score_color = rate_color(overall_rate)

    console.print("[bold]  Scorecard[/bold]\n")
    console.print(
        f"  {score_color(f'{(overall_rate * 100):.0f}%')} overall  "
        f"[dim]({total_passed} passed, {total_failed} failed out of "
        f"{total_assertions} assertions)[/dim]"
    )

    summary = result.get("provider_summary") or result.get("providerSummary") or {}
    for key, stats in summary.items():
        mean_rate = stats.get("pass_rate", {}).get("mean", 0)
        color = rate_color(mean_rate)
        time_mean = stats.get("time_seconds", {}).get("mean", 0)
        tok_mean = stats.get("total_tokens", {}).get("mean", 0)
        cost_mean = stats.get("cost_usd", {}).get("mean", 0)
        console.print(
            f"  {color(f'{(mean_rate * 100):.0f}%')} {key}  "
            f"[dim]{time_mean:.1f}s avg · {tok_mean:.0f} tok · "
            f"${cost_mean:.4f}[/dim]"
        )
    console.print()
